src/freenet/node/diagnostics/diagnostics/threads/ThreadsDiagnostics.java

```diff
@@ -0,0 +1,141 @@
+/* This code is part of Freenet. It is distributed under the GNU General
+ * Public License, version 2 (or at your option any later version). See
+ * http://www.gnu.org/ for further details of the GPL. */
+package freenet.node.diagnostics.diagnostics.threads;
+
+import freenet.node.*;
+import freenet.node.diagnostics.diagnostics.*;
+import freenet.support.*;
+
+import java.lang.management.*;
+import java.util.*;
+
+/**
+ * Runnable thread to retrieve node thread's information and compiling it into
+ * an array of NodeThreadInfo objects.
+ */
+public class ThreadsDiagnostics implements Runnable, Diagnostics {
+    private final String name;
+    private final int monitorInterval;
+
+    private final NodeStats nodeStats;
+    private final Ticker ticker;
+
+    /** Interval in ms the ThreadMonitor is launched */
+    private static final int MONITOR_INTERVAL = 10000;
+    private static final String MONITOR_THREAD_NAME = "NodeDiagnostics: thread monitor";
+
+    /** Sleep interval to calculate % CPU used by each thread */
+    private static final int CPU_SLEEP_INTERVAL = 1000;
+    List<NodeThreadInfo> nodeThreadInfos    = Collections.synchronizedList(new ArrayList<>());
+
+    OperatingSystemMXBean operatingSystemMXBean = ManagementFactory.getOperatingSystemMXBean();
+    ThreadMXBean threadMxBean               = ManagementFactory.getThreadMXBean();
+    RuntimeMXBean runtimeMxBean             = ManagementFactory.getRuntimeMXBean();
+
+    /**
+     * @param nodeStats Used to retrieve data points
+     * @param ticker Used to queue timed jobs
+     * @param name Thread name
+     * @param monitorInterval Sleep intervals to retrieve CPU usage
+     */
+    public ThreadsDiagnostics(NodeStats nodeStats, Ticker ticker, String name, int monitorInterval) {
+        this.nodeStats = nodeStats;
+        this.ticker = ticker;
+        this.name = name;
+        this.monitorInterval = monitorInterval;
+    }
+
+    /*
+     * @param nodeStats Used to retrieve data points
+     * @param ticker Used to queue timed jobs
+     */
+    public ThreadsDiagnostics(NodeStats nodeStats, Ticker ticker) {
+        this(nodeStats, ticker, MONITOR_THREAD_NAME, MONITOR_INTERVAL);
+    }
+
+    private void scheduleNext(int interval) {
+        ticker.queueTimedJob(
+            this,
+            name,
+            interval,
+            false,
+            true
+        );
+    }
+
+    public void start() {
+        scheduleNext(0);
+    }
+
+    private void scheduleNext() {
+        scheduleNext(monitorInterval);
+    }
+
+    @Override
+    public void run() {
```

Comment:
I would suggest something different here, in order to get rid of the `sleep` in here which I still find very distressing. 😀

1. Get snapshot T of all threads and their CPU time.
2. If run for the first time (some `lastRun` is `-1` or whatever), skip until 6.
3. Calculate total CPU time since last run (i.e. time ⨉ CPU count).
4. For all threads in T that we have a previous time for, calculate percentage of runtime.
5. Store all those times in some thread info object.
6. Store T so we can access it in the next invocation.
7. Store current time in `lastRun`.

(From the top of my head, might include mistakes, use with discretion.)

This way the thread running this wouldn’t block unproductively for a whole second (an eternity for a computer 😁).

Reply:
I see where are you going. I'll see if I can make it work.